src/seccomp_utils.c

```diff
@@ -22,25 +22,393 @@
 #include <stdio.h>
 #include <unistd.h>
 #include <string.h>
+#include <search.h>
 #include <ctype.h>
 #include <stdbool.h>
 #include <stdlib.h>
+
+// needed for search mappings
+#include <linux/can.h>
 #include <sys/prctl.h>
+#include <sys/resource.h>
+#include <sys/socket.h>
+#include <sys/types.h>
 
 #include <seccomp.h>
 
-#include "seccomp.h"
+#include "seccomp_utils.h"
 #include "utils.h"
 
+#define sc_map_add(X) sc_map_add_kvp(#X, X)
+
+// libseccomp maximum per ARG_COUNT_MAX in src/arch.h
+#define SC_ARGS_MAXLENGTH	6
 #define SC_MAX_LINE_LENGTH	82	// 80 + '\n' + '\0'
 
 char *filter_profile_dir = "/var/lib/snapd/seccomp/profiles/";
+struct hsearch_data sc_map_htab;
+
+enum parse_ret {
+	PARSE_INVALID_SYSCALL = -2,
+	PARSE_ERROR = -1,
+	PARSE_OK = 0,
+};
 
 struct preprocess {
 	bool unrestricted;
 	bool complain;
 };
 
+/*
+ * arg_cmp contains items of type scmp_arg_cmp (from SCMP_CMP macro) and
+ * length is the number of items in arg_cmp that are active such that if
+ * length is '3' arg_cmp[0], arg_cmp[1] and arg_cmp[2] are used, when length
+ * is '1' only arg_cmp[0] and when length is '0', none are used.
+ */
+struct seccomp_args {
+	int syscall_nr;
+	unsigned int length;
+	struct scmp_arg_cmp arg_cmp[SC_ARGS_MAXLENGTH];
+};
+
+struct sc_map_entry {
+	ENTRY *e;
+	ENTRY *ep;
+	struct sc_map_entry *next;
+};
+
+struct sc_map_list {
+	struct sc_map_entry *list;
+	int count;
+};
+
+struct sc_map_list *sc_map_entries = NULL;
+
+/*
+ * Setup an hsearch map to map strings in the policy (eg, AF_UNIX) to
+ * scmp_datum_t values. Abstract away hsearch implementation behind sc_map_*
+ * functions in case we want to swap this out.
+ *
+ * sc_map_init()		- initialize the hash map via linked list of
+ * 				  of entries
+ * sc_map_add_kvp(key, value)	- create entry from key/value pair and add to
+ * 				  linked list
+ * sc_map_search(s)	- if found, return scmp_datum_t for key, else set errno
+ * sc_map_destroy()	- destroy the hash map and linked list
+ */
+scmp_datum_t sc_map_search(char *s)
+{
+	ENTRY e;
+	ENTRY *ep = NULL;
+	scmp_datum_t val = 0;
+	errno = 0;
+
+	e.key = s;
+	if (hsearch_r(e, FIND, &ep, &sc_map_htab) == 0)
+		die("hsearch_r failed");
+
+	if (ep != NULL) {
+		scmp_datum_t *p = ep->data;
+		val = *p;
+	} else
+		errno = EINVAL;
+
+	return val;
+}
+
+void sc_map_add_kvp(const char *key, scmp_datum_t value)
+{
+	struct sc_map_entry *node;
+	node = (struct sc_map_entry *)malloc(sizeof(struct sc_map_entry));
+	if (node == NULL)
+		die("Out of memory creating sc_map_entries");
+
+	node->e = (ENTRY *) malloc(sizeof(ENTRY));
+	if (node->e == NULL)
+		die("Out of memory creating ENTRY");
+
+	node->e->key = strdup(key);
+	if (node->e->key == NULL)
+		die("Out of memory creating e->key");
+
+	node->e->data = (void *)malloc(sizeof(scmp_datum_t));
+	if (node->e->data == NULL)
+		die("Out of memory creating e->data");
+	*(scmp_datum_t *) node->e->data = value;
```

Comment:
I think there is some confusion on what I'm doing here. Consider:
```
node->e->data = malloc(sizeof(scmp_datum_t *));
if (node->e->data == NULL)
        die("Out of memory creating e->data");
*(scmp_datum_t *) node->e->data = value;
```

What I'm doing is malloc()ing something the size of a `scmp_datum_t` and assigning the address of that in `node->e->data`, which is a void pointer. Then I dereference `node->e->data` and assign it the variable `value`. Since `node->e->data` is a void pointer, the `(scmp_datum_t *)` is used on the left hand sign to say that `node->e->data` is a pointer to `scmp_datum_t`, and the `*` in front says to dereference that pointer. In this manner, we are not assigning a 64 bit type to a small type. `node->e->data = malloc(sizeof(scmp_datum_t *))` is pointer to pointer, and `*(scmp_datum_t *) node->e->data = value` is 64 bit to 64 bit, therefore the sizes are all correct, and we can demonstrate how sc_map_search() later will get the correct value and that the sizes are correct in the attached file. Here are the relevant bits:
```
scmp_datum_t get_val(ENTRY *e) {
    return *(scmp_datum_t *) e->data;
}

int main(int argc, char *argv[]) {
    ...
    scmp_datum_t value = 4294967297; // bigger than 2^32
    ...
    node->e->data = malloc(sizeof(scmp_datum_t *));
    ...
    *(scmp_datum_t *) node->e->data = value;

    printf("sizeof(&value)=\t\t\t\t\t%u\n", sizeof(&value));
    printf("sizeof(value)=\t\t\t\t\t%u\n", sizeof(value));
    printf("value=\t\t\t\t\t\t%llu\n", value);

    printf("sizeof(node->e->data)=\t\t\t\t%u\n", sizeof(node->e->data));
    printf("sizeof(*(scmp_datum_t *) node->e->data)=\t%u\n", sizeof(*(scmp_datum_t *) node->e->data));
    printf("*(scmp_datum_t *)node->e->data=\t\t\t%llu\n", *(scmp_datum_t *)node->e->data);

    printf("get_val(node->e)=\t\t\t\t%llu\n", get_val(node->e));
```
Now run the program on a 32 bit system (eg, an i386 chroot) since it shows the different sizes:
```
$ gcc ./prove-sizes.c && ./a.out
sizeof(&value)=                             4
sizeof(value)=                              8
value=                                      4294967297
sizeof(node->e->data)=                      4
sizeof(*(scmp_datum_t *) node->e->data)=	8
*(scmp_datum_t *)node->e->data=			    4294967297
get_val(node->e)=				            4294967297
```